src/backtrace.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::hyp_layout::{HYP_STACK_BOTTOM, HYP_STACK_TOP};
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+use core::mem::size_of;
+
+#[derive(Copy, Clone)]
+pub enum BTReturnAddress {
+    ReturnAddress(u64),
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+        }
+    }
+}
+
+pub struct BackTrace {
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = HYP_STACK_BOTTOM;
+        let stack_end = HYP_STACK_TOP;
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
```

Comment:
I think the best thing to do is to check whether fp is between HYP_STACK_TOP and HYP_STACK_BOTTOM. If not, disable the backtrace.

This will both fix the problem that is unaddressed right now, the fact that before we switch to the VA stack (or when we are on the overflow stack) this would fail, and it would also make the calculation of the offset below safe.

Reply:
I put a check in, and if it's out of bounds I just return a "NULL" object.